### ISSUE TITLE
Running unit tests in a Windows container

### ISSUE THREAD

**Issue by johlju** [PowerShell/DscResource.Tests#204]
We are getting dangerously close to 60 minutes (in AppVeyor) running tests in the SqlServerDsc resource module, currently it's around 50 minutes. I dont' dare to create more integration tests at this point. I trying to figure out different solutions to solve this, speeding up tests if possible (see issue https://github.com/PowerShell/SqlServerDsc/issues/979). 

One brilliant idea I had (I thought) was to run unit test in a Docker Windows container, while integration tests ran parallel to that in the build worker. Potentially more than one container could run different unit tests, increasing the speed even more. But I started with one container. I did a PoC locally and it looked promising! Up until I ran it in AppVeyor. In AppVeyor it fails with the following error message.

```
Breakpoints cannot be set in the remote session because remote debugging is not supported by the current host.
    + CategoryInfo          : NotImplemented: (:) [Set-PSBreakpoint], PSNotSupportedException
    + FullyQualifiedErrorId : SetPSBreakpoint:RemoteDebuggerNotSupported,Microsoft.PowerShell.Commands.SetPSBreakpointCommand
    + PSComputerName        : 603d281a195a5f2716668cc3cf9a9d3b4c0018f4694edf509da35d52370b7b26
```
https://ci.appveyor.com/project/johlju/sqlserverdsc/build/9.0.174.0

And it seems Pester uses breakpoints for the CodeCoverage part, and apparently I'm in the wrong state when automating this in AppVeyor (works from a PowerShell console against a container).
Pester adds a breakpoint in this function; https://github.com/pester/Pester/blob/312b05f372ca43a35970cfa41874686e033ae387/Functions/Coverage.ps1#L269-L291

Code coverage was one issue I thought would be a mess, but more how I would aggregate the results from both runs. 
Leaving out code coverage would probably solve this, but it does not seems as a good option, I rather like the code coverage.

Anyway, if someone has an idea how this could possibly be done then here is the code (hope I missed something vital); 
https://github.com/johlju/DscResource.Tests/blob/2c30f372222d14d8ddf8d96ae0aab52f4a8e6598/AppVeyor.psm1#L415-L585
And the branch is here; https://github.com/johlju/DscResource.Tests/blob/docker-tests

/cc @PlagueHO @kwirkykat @mbreakey3

**Comment by PlagueHO** [comment]
@johlju - This is awesome that you've made so much progress with running the tests in a Docker container! This really does open up a lot of opportunities for parallelizing our tests.

From what I can tell, remote debugging a container would require the PSRemoting to be enabled in the container. As the container part of the host, I'm guessing that PSRemoting on AppVeyor is disabled. You might still be able to enable it in the container with `Enable-PSRemoting -SkipNetworkProfileCheck -Force`. Have you tried this? I'm guessing Pester doesn't actually need to connect a remote debugging session - if it does then we'd need to open 5985 on the container and map it to a port on the host.

The challenge around how to combine the output from the container with the mainline output is probably solvable but you'd need to be able to pull the CodeCov files over from the container after run as well as combining the verbose logs.

Still, this is awesome stuff @johlju !

**Comment by johlju** [comment]
@PlagueHO My bad. The integration tests does not affect the code coverage, only unit test, so there is no need to compare and merge as I mentioned above. 
My thought was I would be able to see the Pester code coverage report (missed lines) from both integration tests and unit tests, to see how much coverage each did (but the uploaded result would be merged together and uploaded).
But the way we write integration tests now (using `Start-DscConfiguration` and `Get-DscConfiguration`) that scenario can't work. But that just makes it easier code wise, just upload the resulting test result file from the container. And only the container need to have the `-CodeCoverage` parameter when calling Pester, not Pester that runs the integration tests (speeds up somewhat too). 

It would be much simpler if I could just kick of the integration tests in the container, but my integration tests does not support that (but if we get this running we should support any tests to run in a container).
We could use something below to tell which container should run what, `Container = 'Container2'` means all tests that has `Container = 'Container2'` will run in the same container. 

```
[Microsoft.DscResourceKit.IntegrationTest(OrderNumber = 1, Container = 'Container1')]
param()
```
```
[Microsoft.DscResourceKit.UnitTest(Container = 'Container2')]
param()
```

**Comment by johlju** [comment]
@PlagueHO I think it has something to do with the kind of host AppVeyor is running the PowerShell code from, but not sure. This _does work_ when running from a PowerShell console on the build worker (see reproducible steps below). If I output the name of the "host" that the Build Worker is running under it says `RemoteHostImplementation (v1.0.0.0)`, so is it that the script is run in a different kind of "host" than when ran through PowerShell console that give this problem? Maybe we need to turn something on? 

I did ran `Enable-PSRemoting -SkipNetworkProfileCheck -Force`, both on the Build Worker and on the container, but I did not open any ports, so that might still be a problem that I haven't done that yet. Although, I did not need to to this when reproducing the working scenario in the PowerShell console. But maybe `Enable-PSRemoting` plus something else need to be turned on in the "host" on the build worker that run our script. Maybe the port need to be opened in the build worker? 

I checked the value from `$Host.DebuggerEnabled` and it says `$true` for both the Build Worker and the container. Don't know if it has any meaning for this though.

I would love some more tips and ideas here. For example, can we get AppVeyor to start a "normal PowerShell console"-session to run this in? 😄 

**Reproduce working scenario using CodeCoverage in a container ran from PowerShell console**

1. Clone the 'debug repo' https://github.com/johlju/SqlServerDsc/tree/docker-test
1. Push the 'debug repo' to your fork of SqlServerDsc.
1. See AppVeyor test run for for the connection information for RDP to the build worker.
1. Wait for the common tests to start (at that time the container is started and prereq's are installed).
1. In the RDP session of the build worker, in a PowerShell console, run the following. *Note: I could have simplified this code a lot for this example, but wanted to show it reproducing with the actual code that is run.*
    ```
    $containerId = (docker inspect $(docker ps --format "{{.ID}}") | ConvertFrom-Json).Id

    $env:APPVEYOR_BUILD_FOLDER = 'C:\projects\sqlserverdsc'

    $testResultsFile = Join-Path -Path $env:APPVEYOR_BUILD_FOLDER `
                                    -ChildPath 'TestsResults.xml'

    $pesterParameters = @{
        OutputFormat = 'NUnitXML'
        OutputFile   = $testResultsFile
        PassThru     = $True
    }

    if (Test-Path -Path "$env:APPVEYOR_BUILD_FOLDER\DSCResources" )
    {
        $pesterParameters += @{
            CodeCoverage = @(
                "$env:APPVEYOR_BUILD_FOLDER\*.psm1"
                "$env:APPVEYOR_BUILD_FOLDER\DSCResources\*.psm1"
                "$env:APPVEYOR_BUILD_FOLDER\DSCResources\**\*.psm1"
            )
        }
    }

    $getChildItemParameters = @{
        Path = $env:APPVEYOR_BUILD_FOLDER
        Recurse = $true
    }

    $getChildItemParameters['Filter'] = '*.Tests.ps1'
    $testFiles = Get-ChildItem @getChildItemParameters

    $testObjects = @()

    foreach ($testFile in $testFiles)
    {
        $testObjects += @(
            [PSCustomObject] @{
                TestPath = $testFile.FullName
                OrderNumber = $null
            }
        )
    }

    $testObjects | Where-Object -FilterScript {
        $_.TestPath -match 'DSCResource.Tests'
    } | ForEach-Object -Process {
        $_.OrderNumber = 0
    }

    $testObjectOrderContainer = @()

    $testObjectOrderContainer += $testObjects | Where-Object -FilterScript {
        $null -eq $_.OrderNumber `
        -and $_.TestPath -notmatch 'Integration.Tests'
    }

    $pesterParametersContainer = $pesterParameters.Clone()
    $pesterParametersContainer['Path'] = $testObjectOrderContainer.TestPath

    $runPesterAsJobScriptBlock = {
        $pesterParameter = $Using:pesterParametersContainer

        $startJobResult = Start-Job -ScriptBlock {
            param
            (
                # A hash table containing the Pester arguments.
                [Parameter(Mandatory = $true)]
                [System.Collections.Hashtable]
                $PesterParameter
            )

            return Invoke-Pester @PesterParameter
        } -ArgumentList @(
            $pesterParameter
        )

        return $startJobResult
    }

    $containerRemoteSession = New-PSSession -ContainerId $containerId -RunAsAdministrator

    $startJobResult = Invoke-Command `
        -Session $containerRemoteSession `
        -ScriptBlock $runPesterAsJobScriptBlock
    ```
1. The job is running now, which can be shown by running
    ```
    Invoke-Command `
        -Session $containerRemoteSession `
        -ScriptBlock {
            Get-Job -Id $Using:startJobResult.Id
        }
    ```
1. And the result from the job can be fetch by running this (using `-Keep` so this can be run several times, for debug purposes). *Note: It can take a while for it to show output while code coverage is calculated.* 
    ```
    Invoke-Command `
        -Session $containerRemoteSession `
        -ScriptBlock {
            Receive-Job -Id $Using:startJobResult.Id -Keep
        }
    ```

**Comment by johlju** [comment]
Btw. if I turn off code coverage, everything is working as expected (except missing the code coverage); https://ci.appveyor.com/project/johlju/sqlserverdsc/build/9.0.228.0#L1358

**Comment by johlju** [comment]
@PlagueHO I removed the container part in a second branch '[background-tests](https://github.com/johlju/DscResource.Tests/blob/background-tests/AppVeyor.psm1#L421-L473)' and only using a background job (with `Start-Job`). I'm getting the same error when running this. So it seems it has to do with using a background job. See error from run here; https://ci.appveyor.com/project/johlju/sqlserverdsc/build/9.0.235.0#L1356

Appreciate any suggestions? 😄 This is frustrating, I'm _so close_ to get this working. 😆 

The reason for using a background job (in the container) was to be able to get the verbose output from the tests, as well getting the object from Pester (which some logic needs after testing). I'm thinking now that it might not be possible using `Start-Job`. Maybe I should look into starting the container with something like `powershell.exe -Command <script>` (instead of `ping`). Where <script> does the following (top of my head).

- Start transcript against a file
- Start Pester unit test run
- Write Pester result object to a file in JSON format.
- Stop transcript. 
- exit
- container stops 

From the docker container we can extract the transcript file and the JSON file. Transcript file is read and written as output to the "AppVeyor console", this will probably loose colors doing this but at least the output is shown. 
Then the JSON file can be read and converted back to an object that can be used by the logic that follows testing.

I'm I stepping over mountains to do this - is there a simpler way?

**Comment by johlju** [comment]
So I have made a third iteration of this 😆 A PoC using the method above... Finally I have a working testing with coverage! Although the output of the test run is somewhat "ugly", I run the test results thru a [`foreach`-loop which runs `Describe`-, `Context` and `It'-blocks on the test result](https://github.com/johlju/DscResource.Tests/blob/e04aa7587d5c5c38ec0e0564b5c5974d42219466/AppVeyor.psm1#L568-L597). Hopefully that will be sufficient for "verbose output" and failing the test run if any test failed. *I will move the code to functions, but for the PoC I just wanted it to work.* :smile:

I ran this on SqlServerDsc and I'm so happy that it took **20 min less time to finish all tests, that's 40% faster** - if I can start another container splitting the unit tests apart a bit we can improve this even more!

Any suggestions are more than welcome - for this and and any of the other iterations above. :smile: I will continue working on this next weekend.

**Comment by PlagueHO** [comment]
Wow! That is fantastic @johlju - I can't wait to check this out! Really impressive stuff. The other opportunity to use this method with great benefit is for some of the destructive integration tests. For example, some tests can be fairly damaging to the host system, so being able to run them in a disposable environment has a lot of benefits. I'm really looking forward to seeing how this works. I'll be back into this later on in the week!

**Comment by johlju** [comment]
I have everything working now - with a single container. First true test run here (testing CodeCov); https://github.com/PowerShell/SqlServerDsc/pull/1037

I have prepared for using multiple containers, but I will leave that out of the first PR, it's gonna be a lot of code to review anyway. I'm finishing up with some cleanup and localization and after that I will try to write some unit test for the code (I think I can mock Docker). After that I will send in a PR.

After that PR is merged, I will start working on using multiple containers using the concept we started with ordered testing, decorating the tests with an attribute looking something like this.

```
[Microsoft.DscResourceKit.IntegrationTest(OrderNumber = 1, Container = 'Container1')]
param()
```
```
[Microsoft.DscResourceKit.UnitTest(Container = 'Container2')]
param()
```

**Comment by johlju** [comment]
@PlagueHO can I get your input on this one?

For the next iteration of docker support (see previous comment) in the test framework I want to support pulling any image and using any image for a particular unit or integration test. Gonna try to work on this during Easter.

I want to support the following image naming conventions for specifying an image name, these are by default supported by the docker command, so no problem here.

microsoft/windowsservercore
microsoft/windowsservercore:latest
microsoft/windowsservercore:1709

But what shall we do in these scenarios? Are these assumptions correct?

1. If the image name specified, tag or no tag, does not exist locally the image is pulled from Docker Hub.
1. If the image name does not specify a tag, and the image exist locally, then the local image is used.
1. If the image name is specified with a tag, and the image with the tag exist locally, then the local image is used.
1. If the image name is specified with the 'latest' tag, the latest image will _always_ be pulled from Docker Hub, regardless if there is a 'latest' locally. *Or should it behave like any other image name with a tag (see previous).*

I'm curious if there are a scenario where you'd want to pull the image from Docker Hub regardless if it exist locally. Maybe we need to support a `ForcePullImage` in the decoration (see previous comment).

**Comment by PlagueHO** [comment]
Hi @johlju - great summary.

IMHO:
1. Yes. I'd expected latest would be pulled
2. Yes, I think so. But would also assume "latest" local one used (either with latest tag or newest).
3. Yes
4. Latest should pull latest from docker hub IT the latest version is not found locally (regardless of "latest" tag on local image). I think this is the default docker behavior anyway.

I can't think of a reason why we'd want to "Force" pull an image from Docker - because 4 above would generally behave that way. It can't think of an occasion where I've needed to repull in image down from Docker if I already had that exact version (e.g. all the layer hashes matched).

But I'm thinking this might be one of those occasions that we have to try it out for a little bit and tweak as we go along. I'd like to look at moving xNetworking to this model.

**Comment by johlju** [comment]
@PlagueHO Thanks for your input! Yes this probably won't be 100% correct first time around, so we probably will need to tweak it 😄 

Although, I don't think I explained enough on number 4 for you to understand me, I was in a hurry when I wrote that yesterday :) Let me try again.

4. Currently the code only pulls the image if is does not exist locally, regardless of 'latest' tag. If a test want to use `microsoft/windowsservercore:latest`, and that 'latest' tag exist locally it will be used, but that 'latest' might be older than the 'latest' on Docker Hub, in these cases would we want to _always_ pull the latest from Docker Hub to make sure we always have the latest on the build worker?
Would this be the expected behavior?
The test is using the image name `microsoft/windowsservercore`, so it will use the 'latest' on the build worker regardless of the version. Adding the 'latest' tag to the image name, `microsoft/windowsservercore:latest`,  will always try to pull down the latest from Docker Hub regardless of version locally. *If so I could evaluate `$imageName -match ':latest$'` and always call `docker pull` when the evaluation is `$true`.*

**Comment by kingsleyck** [comment]
I was able to succeed at unit testing class-based DSC resources by leveraging the work done for this issue (https://github.com/kingsleyck/PSModulesDsc).

The problem with unit testing class-based DSC is that Pester needs to be executed in a fresh scope in order to execute Mocks reliably. This is documented in the Pester repo: 

https://github.com/pester/Pester/issues/706

By abstracting the unit tests into containers, you solved the scoping issue and effectively made testing of class-based resources in the HQRM format possible.

**Comment by johlju** [comment]
@kingsleyck That is awesome that you had use for this! Since you have both unit test and integration tests, if you didn't run the unit tests in the containers the integration test would fail with the error you mention?

**Comment by kingsleyck** [comment]
The unit tests fail when Mocks are used on Powershell version 5. I don't think the integration tests failed because they were not using Mocks, but I could be wrong. This issue was also discussed previously for the repository: https://github.com/bgelens/DockerDsc ( thanks @bgelens ). I can't find the discussion now unfortunately. If you examine the AppVeyor output on the repo, though, the Mocking issue is visible.

The typical workaround is to use "powershell.exe -Command Invoke-Pester" or Start-Job which creates a fresh scope for each test run. With a container per unit test the scope is always fresh. I tried various modifications to Invoke-AppveyorTestScriptTask and trying to merge code coverage files before realizing that by containerizing the tests, you had already solved the problem.

The underlying issue with classes not refreshing in scope has been resolved in Powershell 6.1, but that version is not fully supported by the tools in this repository yet. For now using containers is not only faster, it is a functional workaround to the issues testing class-based DSC resources.
